examples/components/CCFRAUD_VERTICAL/traininsilo/contributor.py

```diff
@@ -0,0 +1,404 @@
+        )
+        client.log_param(
+            run_id=run_id,
+            key=f"optimizer {self._experiment_name}",
+            value=self.optimizer_.__class__.__name__,
+        )
+
+    def log_metrics(self, client, run_id, key, value, pipeline_level=False):
+        if pipeline_level:
+            client.log_metric(
+                run_id=run_id,
+                key=f"{self._experiment_name}/{key}",
+                value=value,
+            )
+        else:
+            client.log_metric(
+                run_id=run_id,
+                key=f"{self._iteration_name}/{self._experiment_name}/{key}",
+                value=value,
+            )
+
+    def local_train(self, checkpoint):
+        """Perform local training for a given number of epochs
+
+        Args:
+            checkpoint: Previous model checkpoint from where training has to be started.
+        """
+
+        if checkpoint:
+            self.model_.load_state_dict(torch.load(checkpoint + "/model.pt"))
+
+        with mlflow.start_run() as mlflow_run:
+            # get Mlflow client and root run id
+            mlflow_client = mlflow.tracking.client.MlflowClient()
+            root_run_id = os.environ.get("AZUREML_ROOT_RUN_ID")
+            logger.debug(f"Root runId: {root_run_id}")
+
+            # log params
+            self.log_params(mlflow_client, root_run_id)
+
+            logger.debug("Local training started")
+
+            train_metrics = RunningMetrics(
+                ["loss", "accuracy", "precision", "recall"], prefix="train"
+            )
+
+            for _ in range(1, self._epochs + 1):
+                self.model_.train()
+                train_metrics.reset_global()
+
+                for i, batch in enumerate(self.train_loader_):
+                    data = batch.to(self.device_)
+                    # Zero gradients for every batch
+                    self.optimizer_.zero_grad()
+
+                    output, net_loss = self.model_(data)
+                    output = output.contiguous()
+                    self._global_comm.send(output, 0)
+                    if net_loss is not None:
+                        net_loss += net_loss * 1e-5
+                        net_loss.backward(retain_graph=True)
+                    gradient = self._global_comm.recv(0).to(self.device_)
+                    output.backward(gradient)
+                    self.optimizer_.step()
+
+                self.test()
+
+            log_message = [
+                f"End of training",
+            ]
+            logger.info(", ".join(log_message))
+
+    def test(self):
+        """Test the trained model and report test loss and accuracy"""
+        self.model_.eval()
+
+        with torch.no_grad():
+            for batch in self.test_loader_:
+                data = batch.to(self.device_)
+
+                if self._global_rank != 0:
+                    output = self.model_(data)[0].contiguous()
+                    self._global_comm.send(output, 0)
```

Comment:
can you add a comment there to explain what's happening here?

i.e. : the test just consists in creating outputs, sent to the host for the host to run the last forward layers and compute accuracy/metrics

Reply:
(I initially got confused)